File: windows/src/HajatManager/Services/Exporter.cs
using ClosedXML.Excel;
using HajatManager.Models;
using QuestPDF.Fluent;
using QuestPDF.Helpers;
using QuestPDF.Infrastructure;

namespace HajatManager.Services;

// Export PDF/Excel offline dari SQLite — cermin exporter.dart + web EventClient.
// Header hijau #059669 + baris TOTAL + Rincian per Metode + TTD 3 kolom.
public static class Exporter
{
    public static string FormatRp(long v) => $"Rp {v:N0}".Replace(",", ".");

    public static void ExportPdf(string path, EventModel ev, List<GuestModel> guests, string userName = "-", string? mejaLabel = null)
    {
        QuestPDF.Settings.License = LicenseType.Community;
        var total = guests.Sum(g => g.Nominal);
        var safeEvent = string.IsNullOrWhiteSpace(ev.NamaAcara) ? "Acara" : ev.NamaAcara.Trim();
        var safeUser = string.IsNullOrWhiteSpace(userName) ? "-" : userName.Trim();
        var title = $"{safeEvent} - Laporan Pemberian";
        var meta = $"{DateTime.Now:dd/MM/yyyy HH:mm} | {safeUser} | {guests.Count} tamu | {FormatRp(total)}";
        var byMetode = guests.GroupBy(g => string.IsNullOrEmpty(g.Metode) ? "-" : g.Metode)
            .Select(g => new { Metode = g.Key, Jumlah = g.Count(), Total = g.Sum(x => x.Nominal) })
            .OrderByDescending(x => x.Total).ToList();
        Document.Create(c =>
        {
            c.Page(p =>
            {
                p.Size(PageSizes.A4);
                p.Margin(36);
                p.Header().Column(col =>
                {
                    col.Item().Text(title).FontSize(11).Bold();
                    col.Item().Text(meta).FontSize(8).FontColor(Colors.Grey.Darken2);
                });
                p.Content().Column(col =>
                {
                    col.Item().Table(t =>
                    {
                        t.ColumnsDefinition(d =>
                        {
                            d.ConstantColumn(28);
                            d.RelativeColumn(3);
                            d.RelativeColumn(2);
                            d.ConstantColumn(64);
                            d.ConstantColumn(52);
                            d.ConstantColumn(44);
                            d.RelativeColumn(2);
                        });
                        t.Header(h =>
                        {
                            foreach (var head in new[] { "No", "Nama", "Alamat", "Nominal", "Metode", "Meja", "Catatan" })
                                h.Cell().Background("#059669").Padding(6).Text(head).FontColor(Colors.White).Bold().FontSize(8);
                        });
                        var i = 0;
                        foreach (var g in guests)
                        {
                            i++;
                            t.Cell().Padding(5).Text(i.ToString()).FontSize(7);
                            t.Cell().Padding(5).Text(g.Nama).FontSize(7);
                            t.Cell().Padding(5).Text(g.Alamat).FontSize(7);
                            t.Cell().Padding(5).AlignRight().Text(FormatRp(g.Nominal)).FontSize(7);
                            t.Cell().Padding(5).Text(g.Metode ?? "-").FontSize(7);
                            t.Cell().Padding(5).Text(string.IsNullOrEmpty(g.MejaLabel) ? "-" : g.MejaLabel).FontSize(7);
                            t.Cell().Padding(5).Text(string.IsNullOrEmpty(g.Catatan) ? "-" : g.Catatan).FontSize(7);
                        }
                        t.Cell().Padding(6).Text("").FontSize(8);
                        t.Cell().Padding(6).Text("").FontSize(8);
                        t.Cell().Padding(6).Text("TOTAL").Bold().FontSize(8);
                        t.Cell().Padding(6).AlignRight().Text(FormatRp(total)).Bold().FontSize(8);
                        t.Cell().Padding(6).Text("").FontSize(8);
                        t.Cell().Padding(6).Text($"{guests.Count} tamu").Bold().FontSize(8);
                        t.Cell().Padding(6).Text("").FontSize(8);
                    });
                    // Rincian per Metode
                    col.Item().PaddingTop(12).Text("Rincian per Metode").FontSize(11).Bold();
                    col.Item().Text($"Total {guests.Count} catatan | {FormatRp(total)}").FontSize(7);
                    col.Item().Table(t =>
                    {
                        t.ColumnsDefinition(d =>
                        {
                            d.RelativeColumn(3);
                            d.ConstantColumn(60);
                            d.ConstantColumn(90);
                            d.ConstantColumn(60);
                        });
                        t.Header(h =>
                        {
                            foreach (var head in new[] { "Metode", "Jumlah", "Total", "Porsi" })
                                h.Cell().Background("#059669").Padding(6).Text(head).FontColor(Colors.White).Bold().FontSize(8);
                        });
                        foreach (var m in byMetode)
                        {
                            var porsi = total > 0 ? $"{(int)Math.Round(m.Total * 100.0 / total)}%" : "0%";
                            t.Cell().Padding(5).Text(m.Metode).FontSize(7);
                            t.Cell().Padding(5).AlignCenter().Text(m.Jumlah.ToString()).FontSize(7);
                            t.Cell().Padding(5).AlignRight().Text(FormatRp(m.Total)).FontSize(7);
                            t.Cell().Padding(5).AlignCenter().Text(porsi).FontSize(7);
                        }
                        t.Cell().Padding(6).Text("TOTAL").Bold().FontSize(8);
                        t.Cell().Padding(6).AlignCenter().Text(guests.Count.ToString()).Bold().FontSize(8);
                        t.Cell().Padding(6).AlignRight().Text(FormatRp(total)).Bold().FontSize(8);
                        t.Cell().Padding(6).AlignCenter().Text("100%").Bold().FontSize(8);
                    });
                    // Serah terima 3 kolom
                    col.Item().PaddingTop(12).LineHorizontal(0.5f).LineColor(Colors.Grey.Lighten2);
                    col.Item().Row(r =>
                    {
                        SignBlock(r.RelativeItem(), "Diserahkan oleh,", "Petugas/Admin", false);
                        SignBlock(r.RelativeItem(), "Diterima oleh,", "Owner/Tuan Rumah", false);
                        SignBlock(r.RelativeItem(), "Saksi,", "Nama (opsional)", true);
                    });
                });
                p.Footer().AlignRight().Text(t =>
                {
                    t.CurrentPageNumber().FontSize(7);
                    t.Span(" / ").FontSize(7);
                    t.TotalPages().FontSize(7);
                });
            });
        }).GeneratePdf(path);
    }

    private static void SignBlock(IContainer item, string title, string subtitle, bool optional)
    {
        item.Column(c =>
        {
            c.Item().Text(title).FontSize(7.5f).Bold();
            c.Item().Text(subtitle).FontSize(6.5f).Italic(optional);
            c.Item().PaddingTop(4);
            foreach (var label in new[] { "Nama", "Tanggal", "Tanda tangan" })
            {
                c.Item().Row(r =>
                {
                    r.ConstantItem(52).Text(label).FontSize(7);
                    r.RelativeItem().PaddingBottom(14).BorderBottom(0.5f).BorderColor(Colors.Grey.Medium);
                });
                c.Item().PaddingBottom(label == "Tanda tangan" ? 14 : 8);
            }
        });
    }

    public static void ExportXlsx(string path, EventModel ev, List<GuestModel> guests)
    {
        using var wb = new XLWorkbook();
        var ws = wb.Worksheets.Add("Rekap");
        var safeEvent = string.IsNullOrWhiteSpace(ev.NamaAcara) ? "Acara" : ev.NamaAcara.Trim();
        ws.Cell(1, 1).Value = safeEvent;
        ws.Cell(2, 1).Value = $"Laporan Pemberian | {DateTime.Now:dd/MM/yyyy HH:mm} | {guests.Count} tamu";
        string[] head = { "No", "Nama", "Alamat", "Nominal", "Metode", "Meja", "Catatan" };
        for (var i = 0; i < head.Length; i++)
        {
            var c = ws.Cell(4, i + 1);
            c.Value = head[i];
            c.Style.Font.Bold = true;
            c.Style.Fill.BackgroundColor = XLColor.FromHtml("#059669");
            c.Style.Font.FontColor = XLColor.White;
        }
        var r = 5;
        var no = 0;
        foreach (var g in guests)
        {
            no++;
            ws.Cell(r, 1).Value = no;
            ws.Cell(r, 2).Value = g.Nama;
            ws.Cell(r, 3).Value = g.Alamat;
            ws.Cell(r, 4).Value = g.Nominal;
            ws.Cell(r, 5).Value = g.Metode;
            ws.Cell(r, 6).Value = string.IsNullOrEmpty(g.MejaLabel) ? "-" : g.MejaLabel;
            ws.Cell(r, 7).Value = g.Catatan ?? "-";
            r++;
        }
        ws.Cell(r, 2).Value = "TOTAL";
        ws.Cell(r, 2).Style.Font.Bold = true;
        ws.Cell(r, 4).Value = guests.Sum(g => g.Nominal);
        ws.Cell(r, 4).Style.Font.Bold = true;
        ws.Cell(r, 6).Value = $"{guests.Count} tamu";
        ws.Cell(r, 6).Style.Font.Bold = true;
        r++;
        // Rincian per Metode
        r++;
        ws.Cell(r, 1).Value = "Rincian per Metode";
        ws.Cell(r, 1).Style.Font.Bold = true;
        r++;
        foreach (var h in new[] { "Metode", "Jumlah", "Total", "Porsi%" })
        {
            var c = ws.Cell(r, Array.IndexOf(new[] { "Metode", "Jumlah", "Total", "Porsi%" }, h) + 1);
            c.Value = h;
            c.Style.Font.Bold = true;
            c.Style.Fill.BackgroundColor = XLColor.FromHtml("#059669");
            c.Style.Font.FontColor = XLColor.White;
        }
        r++;
        var total = guests.Sum(g => g.Nominal);
        foreach (var m in guests.GroupBy(g => string.IsNullOrEmpty(g.Metode) ? "-" : g.Metode)
            .Select(g => new { Metode = g.Key, Jumlah = g.Count(), Total = g.Sum(x => x.Nominal) })
            .OrderByDescending(x => x.Total))
        {
            ws.Cell(r, 1).Value = m.Metode;
            ws.Cell(r, 2).Value = m.Jumlah;
            ws.Cell(r, 3).Value = m.Total;
            ws.Cell(r, 4).Value = total > 0 ? $"{(int)Math.Round(m.Total * 100.0 / total)}%" : "0%";
            r++;
        }
        ws.Cell(r, 1).Value = "TOTAL";
        ws.Cell(r, 1).Style.Font.Bold = true;
        ws.Cell(r, 2).Value = guests.Count;
        ws.Cell(r, 2).Style.Font.Bold = true;
        ws.Cell(r, 3).Value = total;
        ws.Cell(r, 3).Style.Font.Bold = true;
        ws.Cell(r, 4).Value = "100%";
        ws.Cell(r, 4).Style.Font.Bold = true;
        ws.Columns().AdjustToContents();
        wb.SaveAs(path);
    }
}

File: mobile/lib/core/exporter.dart
import 'dart:io';
import 'dart:typed_data';
import 'package:excel/excel.dart';
import 'package:intl/date_symbol_data_local.dart';
import 'package:intl/intl.dart';
import 'package:path_provider/path_provider.dart';
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;
import 'package:share_plus/share_plus.dart';
import 'format_rp.dart';
import 'local_db.dart';

/// Export mirror web (`EventClient.handleExport`):
/// jenis Pemberian/Buku Tamu, urutan, orientasi, header hijau + footer TOTAL.
class ExportOptions {
  final String type; // 'pemberian' | 'tamu'
  final String order; // nama_az|nama_za|alamat_az|nominal_desc|waktu_desc
  final bool landscape;
  const ExportOptions(
      {this.type = 'pemberian',
      this.order = 'waktu_desc',
      this.landscape = false});
}

const _exportOrders = {
  'waktu_desc': 'Waktu terbaru',
  'nama_az': 'Nama A–Z',
  'nama_za': 'Nama Z–A',
  'alamat_az': 'Alamat A–Z',
  'nominal_desc': 'Nominal terbesar',
};

Map<String, String> exportOrderLabels() =>
    Map<String, String>.from(_exportOrders);

List<Map<String, dynamic>> sortGuests(
    List<Map<String, dynamic>> rows, String order) {
  final list = [...rows];
  int cmpStr(String a, String b) =>
      a.toLowerCase().compareTo(b.toLowerCase());
  switch (order) {
    case 'nama_az':
      list.sort((a, b) => cmpStr('${a['nama']}', '${b['nama']}'));
    case 'nama_za':
      list.sort((a, b) => cmpStr('${b['nama']}', '${a['nama']}'));
    case 'alamat_az':
      list.sort((a, b) {
        final c = cmpStr('${a['alamat']}', '${b['alamat']}');
        return c != 0 ? c : cmpStr('${a['nama']}', '${b['nama']}');
      });
    case 'nominal_desc':
      list.sort((a, b) => ((b['nominal'] as int?) ?? 0)
          .compareTo((a['nominal'] as int?) ?? 0));
    case 'waktu_desc':
    default:
      list.sort((a, b) => '${b['createdAt'] ?? ''}'
          .compareTo('${a['createdAt'] ?? ''}'));
  }
  return list;
}

List<Map<String, dynamic>> sortBooks(
    List<Map<String, dynamic>> rows, String order) {
  final list = [...rows];
  int cmpStr(String a, String b) =>
      a.toLowerCase().compareTo(b.toLowerCase());
  switch (order) {
    case 'nama_za':
      list.sort((a, b) => cmpStr('${b['nama']}', '${a['nama']}'));
    case 'alamat_az':
      list.sort((a, b) {
        final c = cmpStr('${a['alamat']}', '${b['alamat']}');
        return c != 0 ? c : cmpStr('${a['nama']}', '${b['nama']}');
      });
    case 'nama_az':
    default:
      list.sort((a, b) => cmpStr('${a['nama']}', '${b['nama']}'));
  }
  return list;
}

final _green = PdfColor.fromInt(0xFF059669);
final _greenLight = PdfColor.fromInt(0xFFD1FAE5);
final _greenDark = PdfColor.fromInt(0xFF064E3B);
final _greyLine = PdfColor.fromInt(0xFFD2D2D2);
final _greyAlt = PdfColor.fromInt(0xFFF9FAFB);

bool _localeReady = false;

Future<void> ensureLocaleId() async {
  if (_localeReady) return;
  try {
    await initializeDateFormatting('id_ID', null);
    _localeReady = true;
  } catch (_) {
    // Fallback: pakai locale default, jangan throw saat export.
  }
}

String _stamp() {
  try {
    return DateFormat('d/M/yyyy HH:mm', 'id_ID').format(DateTime.now());
  } catch (_) {
    // Belum init (misal path export dipanggil sebelum main selesai):
    // fallback tanpa locale agar export tetap jalan.
    final n = DateTime.now();
    final d = n.day.toString().padLeft(2, '0');
    final m = n.month.toString().padLeft(2, '0');
    final h = n.hour.toString().padLeft(2, '0');
    final mi = n.minute.toString().padLeft(2, '0');
    return '$d/$m/${n.year} $h:$mi';
  }
}

String _fileDate() => DateFormat('yyyy-MM-dd').format(DateTime.now());

pw.Widget _cell(String t,
    {bool bold = false,
    PdfColor? color,
    double fontSize = 7,
    pw.Alignment align = pw.Alignment.centerLeft}) =>
    pw.Container(
      alignment: align,
      padding: const pw.EdgeInsets.all(4),
      child: pw.Text(t,
          style: pw.TextStyle(
              fontSize: fontSize,
              fontWeight:
                  bold ? pw.FontWeight.bold : pw.FontWeight.normal,
              color: color)),
    );

Future<Uint8List> buildExportPdf({
  required String eventName,
  required String userName,
  required ExportOptions opt,
  required List<Map<String, dynamic>> rows,
}) async {
  await ensureLocaleId();
  final doc = pw.Document();
  final landscaped = opt.landscape;
  final format =
      landscaped ? PdfPageFormat.a4.landscape : PdfPageFormat.a4;
  final isPemberian = opt.type == 'pemberian';
  final headers = isPemberian
      ? ['No', 'Nama', 'Alamat', 'Nominal', 'Metode', 'Meja', 'Catatan']
      : ['No', 'Nama', 'Alamat'];
  final widths = isPemberian
      ? <int, pw.TableColumnWidth>{
          0: const pw.FixedColumnWidth(24),
          3: const pw.FixedColumnWidth(64),
          4: const pw.FixedColumnWidth(52),
          5: const pw.FixedColumnWidth(44),
        }
      : <int, pw.TableColumnWidth>{
          0: const pw.FixedColumnWidth(28),
        };
  int totalNominal = 0;
  final bodyRows = <pw.TableRow>[];
  for (var i = 0; i < rows.length; i++) {
    final g = rows[i];
    final cells = isPemberian
        ? [
            '${i + 1}',
            '${g['nama']}',
            '${g['alamat']}',
            formatRp((g['nominal'] as int?) ?? 0),
            '${g['metode'] ?? '-'}',
            '${g['mejaLabel'] ?? '-'}',
            '${g['catatan'] ?? '-'}',
          ]
        : ['${i + 1}', '${g['nama']}', '${g['alamat']}'];
    if (isPemberian) totalNominal += (g['nominal'] as int?) ?? 0;
    bodyRows.add(pw.TableRow(
      decoration: pw.BoxDecoration(
          color: i.isEven ? PdfColors.white : _greyAlt),
      children: cells.map((t) => _cell(t)).toList(),
    ));
  }
  final footCells = isPemberian
      ? ['', '', 'TOTAL', formatRp(totalNominal), '',
          '${rows.length} tamu', '']
      : ['', 'TOTAL', '${rows.length} tamu'];
  // Title aman-font: ASCII saja (| dan -), jangan samakan web yang pakai
  // em-dash/bullet (U+2014/U+2022) karena Helvetica standar pdf-Dart throw.
  final safeEvent =
      eventName.trim().isEmpty ? 'Acara' : eventName.trim();
  final safeUser = userName.trim().isEmpty ? '-' : userName.trim();
  final title =
      '$safeEvent - ${isPemberian ? 'Laporan Pemberian' : 'Buku Tamu'}';
  final meta =
      '${_stamp()} | $safeUser | ${rows.length} tamu${isPemberian ? ' | ${formatRp(totalNominal)}' : ''}';
  final perPage = landscaped ? 24 : 32;
  final totalPages = (rows.length / perPage).ceil().clamp(1, 9999);
  pw.Widget table() => pw.Column(children: [
        pw.Table(
          border: pw.TableBorder.all(color: _greyLine, width: 0.4),
          columnWidths: widths,
          children: [
            pw.TableRow(
              decoration: pw.BoxDecoration(color: _green),
              children: headers
                  .map((h) => _cell(h,
                      bold: true,
                      color: PdfColors.white,
                      fontSize: 8,
                      align: pw.Alignment.center))
                  .toList(),
            ),
            ...bodyRows,
            pw.TableRow(
              decoration: pw.BoxDecoration(color: _greenLight),
              children: footCells
                  .map((t) => _cell(t,
                      bold: true,
                      color: _greenDark,
                      fontSize: 8,
                      align: pw.Alignment.center))
                  .toList(),
            ),
          ],
        ),
      ]);

  // --- Rincian per Metode (cermin web EventClient 1213-1244) ---
  pw.Widget metodeTable() {
    if (!isPemberian) return pw.SizedBox();
    final map = <String, (int, int)>{};
    for (final g in rows) {
      final m = '${g['metode'] ?? '-'}';
      final n = (g['nominal'] as int?) ?? 0;
      final cur = map[m] ?? (0, 0);
      map[m] = (cur.$1 + 1, cur.$2 + n);
    }
    final entries = map.entries.toList()
      ..sort((a, b) => b.value.$2.compareTo(a.value.$2));
    final mBody = <pw.TableRow>[];
    for (var i = 0; i < entries.length; i++) {
      final e = entries[i];
      final porsi = totalNominal > 0
          ? '${((e.value.$2 / totalNominal) * 100).round()}%'
          : '0%';
      mBody.add(pw.TableRow(
        decoration: pw.BoxDecoration(
            color: i.isEven ? PdfColors.white : _greyAlt),
        children: [
          _cell(e.key),
          _cell('${e.value.$1}', align: pw.Alignment.center),
          _cell(formatRp(e.value.$2), align: pw.Alignment.centerRight),
          _cell(porsi, align: pw.Alignment.center),
        ],
      ));
    }
    return pw.Column(
      crossAxisAlignment: pw.CrossAxisAlignment.start,
      children: [
        pw.SizedBox(height: 12),
        pw.Text('Rincian per Metode',
            style:
                pw.TextStyle(fontSize: 11, fontWeight: pw.FontWeight.bold)),
        pw.Text('Total ${rows.length} catatan | ${formatRp(totalNominal)}',
            style: const pw.TextStyle(fontSize: 7)),
        pw.SizedBox(height: 6),
        pw.Table(
          border: pw.TableBorder.all(color: _greyLine, width: 0.4),
          children: [
            pw.TableRow(
              decoration: pw.BoxDecoration(color: _green),
              children: ['Metode', 'Jumlah', 'Total', 'Porsi']
                  .map((h) => _cell(h,
                      bold: true,
                      color: PdfColors.white,
                      fontSize: 8,
                      align: pw.Alignment.center))
                  .toList(),
            ),
            ...mBody,
            pw.TableRow(
              decoration: pw.BoxDecoration(color: _greenLight),
              children: [
                _cell('TOTAL', bold: true, color: _greenDark, fontSize: 8,
                    align: pw.Alignment.center),
                _cell('${rows.length}', bold: true, color: _greenDark,
                    fontSize: 8, align: pw.Alignment.center),
                _cell(formatRp(totalNominal), bold: true, color: _greenDark,
                    fontSize: 8, align: pw.Alignment.centerRight),
                _cell('100%', bold: true, color: _greenDark, fontSize: 8,
                    align: pw.Alignment.center),
              ],
            ),
          ],
        ),
      ],
    );
  }

  // --- Serah terima 3 kolom (cermin web 1246-1294) ---
  pw.Widget signBlock(String title, String subtitle, bool optional) =>
      pw.Expanded(
        child: pw.Column(
          crossAxisAlignment: pw.CrossAxisAlignment.start,
          children: [
            pw.Text(title,
                style: pw.TextStyle(
                    fontSize: 7.5, fontWeight: pw.FontWeight.bold)),
            pw.Text(subtitle,
                style: pw.TextStyle(
                    fontSize: 6.5,
                    fontStyle: optional
                        ? pw.FontStyle.italic
                        : pw.FontStyle.normal)),
            pw.SizedBox(height: 4),
            for (final label in ['Nama', 'Tanggal', 'Tanda tangan'])
              pw.Column(
                crossAxisAlignment: pw.CrossAxisAlignment.start,
                children: [
                  pw.Row(children: [
                    pw.SizedBox(
                        width: 52, child: _cell(label, fontSize: 7)),
                    pw.Expanded(
                        child: pw.Container(
                            height: 14,
                            decoration: const pw.BoxDecoration(
                                border: pw.Border(
                                    bottom: pw.BorderSide(
                                        color: PdfColors.grey, width: 0.5))))),
                  ]),
                  pw.SizedBox(height: label == 'Tanda tangan' ? 14 : 8),
                ],
              ),
          ],
        ),
      );

  pw.Widget signSection() {
    if (!isPemberian) return pw.SizedBox();
    return pw.Column(children: [
      pw.SizedBox(height: 12),
      pw.Divider(color: _greyLine, thickness: 0.5),
      pw.SizedBox(height: 6),
      pw.Row(
        crossAxisAlignment: pw.CrossAxisAlignment.start,
        children: [
          signBlock('Diserahkan oleh,', 'Petugas/Admin', false),
          pw.SizedBox(width: 12),
          signBlock('Diterima oleh,', 'Owner/Tuan Rumah', false),
          pw.SizedBox(width: 12),
          signBlock('Saksi,', 'Nama (opsional)', true),
        ],
      ),
    ]);
  }

  doc.addPage(pw.MultiPage(
    pageFormat: format,
    margin: const pw.EdgeInsets.all(16),
    header: (ctx) => ctx.pageNumber > 1
        ? pw.SizedBox()
        : pw.Column(
            crossAxisAlignment: pw.CrossAxisAlignment.start,
            children: [
                pw.Text(title,
                    maxLines: 2,
                    style: pw.TextStyle(
                        fontSize: 11,
                        fontWeight: pw.FontWeight.bold)),
                pw.Text(meta,
                    style: const pw.TextStyle(fontSize: 8)),
                pw.SizedBox(height: 8),
              ]),
    footer: (ctx) => pw.Container(
        alignment: pw.Alignment.centerRight,
        child: pw.Text(
            'Halaman ${ctx.pageNumber} / $totalPages',
            style: const pw.TextStyle(fontSize: 7))),
    build: (_) => [table(), metodeTable(), signSection()],
  ));
  return doc.save();
}

void _styleLastRow(Sheet sh, CellStyle style) {
  if (sh.rows.isEmpty) return;
  for (final cell in sh.rows.last) {
    cell?.cellStyle = style;
  }
}

List<int>? buildExportXlsx({
  required String eventName,
  required ExportOptions opt,
  required List<Map<String, dynamic>> rows,
}) {
  final isPemberian = opt.type == 'pemberian';
  final ex = Excel.createExcel();
  final sh = ex['Rekap'];
  final headStyle = CellStyle(
      bold: true,
      backgroundColorHex: ExcelColor.fromHexString('#059669'),
      fontColorHex: ExcelColor.fromHexString('#FFFFFF'));
  final footStyle = CellStyle(
      bold: true,
      backgroundColorHex: ExcelColor.fromHexString('#D1FAE5'),
      fontColorHex: ExcelColor.fromHexString('#064E3B'));
  final safeEventX =
      eventName.trim().isEmpty ? 'Acara' : eventName.trim();
  sh.appendRow([TextCellValue(safeEventX)]);
  sh.appendRow([
    TextCellValue(
        '${isPemberian ? 'Laporan Pemberian' : 'Buku Tamu'} | ${_stamp()} | ${rows.length} tamu')
  ]);
  sh.appendRow([]);
  final head = isPemberian
      ? ['No', 'Nama', 'Alamat', 'Nominal', 'Metode', 'Meja', 'Catatan']
      : ['No', 'Nama', 'Alamat'];
  sh.appendRow(head.map(TextCellValue.new).toList());
  _styleLastRow(sh, headStyle);
  int totalNominal = 0;
  for (var i = 0; i < rows.length; i++) {
    final g = rows[i];
    if (isPemberian) {
      totalNominal += (g['nominal'] as int?) ?? 0;
      sh.appendRow([
        IntCellValue(i + 1),
        TextCellValue('${g['nama']}'),
        TextCellValue('${g['alamat']}'),
        IntCellValue((g['nominal'] as int?) ?? 0),
        TextCellValue('${g['metode'] ?? '-'}'),
        TextCellValue('${g['mejaLabel'] ?? '-'}'),
        TextCellValue('${g['catatan'] ?? '-'}'),
      ]);
    } else {
      sh.appendRow([
        IntCellValue(i + 1),
        TextCellValue('${g['nama']}'),
        TextCellValue('${g['alamat']}'),
      ]);
    }
  }
  if (isPemberian) {
    sh.appendRow([
      TextCellValue(''),
      TextCellValue(''),
      TextCellValue('TOTAL'),
      IntCellValue(totalNominal),
      TextCellValue(''),
      TextCellValue('${rows.length} tamu'),
      TextCellValue(''),
    ]);
  } else {
    sh.appendRow([
      TextCellValue(''),
      TextCellValue('TOTAL'),
      TextCellValue('${rows.length} tamu'),
    ]);
  }
  _styleLastRow(sh, footStyle);
  // Section Rincian per Metode (cermin PDF/web).
  if (isPemberian) {
    final map = <String, (int, int)>{};
    for (final g in rows) {
      final m = '${g['metode'] ?? '-'}';
      final n = (g['nominal'] as int?) ?? 0;
      final cur = map[m] ?? (0, 0);
      map[m] = (cur.$1 + 1, cur.$2 + n);
    }
    final entries = map.entries.toList()
      ..sort((a, b) => b.value.$2.compareTo(a.value.$2));
    sh.appendRow([]);
    sh.appendRow([TextCellValue('Rincian per Metode')]);
    sh.appendRow(['Metode', 'Jumlah', 'Total', 'Porsi%']
        .map(TextCellValue.new)
        .toList());
    _styleLastRow(sh, headStyle);
    for (final e in entries) {
      final porsi = totalNominal > 0
          ? '${((e.value.$2 / totalNominal) * 100).round()}%'
          : '0%';
      sh.appendRow([
        TextCellValue(e.key),
        IntCellValue(e.value.$1),
        IntCellValue(e.value.$2),
        TextCellValue(porsi),
      ]);
    }
    sh.appendRow([
      TextCellValue('TOTAL'),
      IntCellValue(rows.length),
      IntCellValue(totalNominal),
      TextCellValue('100%'),
    ]);
    _styleLastRow(sh, footStyle);
  }
  for (var i = 0; i < head.length; i++) {
    sh.setColumnWidth(i, isPemberian ? 18 : 24);
  }
  return ex.save();
}

/// Ambil baris lokal sesuai opsi (offline OK).
Future<List<Map<String, dynamic>>> exportRows(
    String eventId, ExportOptions opt) async {
  if (opt.type == 'tamu') {
    final books = await LocalDb.instance.booksLocal(eventId);
    return sortBooks(books, opt.order);
  }
  final guests = await LocalDb.instance.guestsLocal(eventId);
  return sortGuests(guests, opt.order);
}

/// Simpan ke file temp + share. Fallback: kalau Share gagal (misal Linux
/// tanpa handler / dbus), simpan ke Documents dan kembalikan path agar UI
/// bisa tampilkan "tersimpan di ..." bukan cuma "Export gagal".
Future<String> shareExportFile(
    List<int> bytes, String filename, String mime) async {
  final dir = await getTemporaryDirectory();
  final f = await File('${dir.path}/$filename')
      .writeAsBytes(bytes, flush: true);
  try {
    await SharePlus.instance.share(ShareParams(
      files: [XFile(f.path, mimeType: mime)],
      subject: filename,
    ));
    return f.path;
  } catch (_) {
    try {
      final docs = await getApplicationDocumentsDirectory();
      final g = await File('${docs.path}/$filename')
          .writeAsBytes(bytes, flush: true);
      return g.path;
    } catch (_) {
      // Terakhir: kembalikan path temp agar pemanggil tetap info lokasi.
      return f.path;
    }
  }
}

String exportFilename(String eventName, ExportOptions opt, String ext) {
  var safe = eventName
      .replaceAll(RegExp(r'[^\w\- ]+'), '')
      .trim()
      .replaceAll(RegExp(r'\s+'), '-')
      .toLowerCase();
  if (safe.isEmpty) safe = 'acara';
  if (safe == '-$ext' || safe.startsWith('-')) safe = 'acara$safe';
  final kind = opt.type == 'pemberian' ? 'pemberian' : 'buku-tamu';
  return '$safe-$kind-${_fileDate()}.$ext';
}
